L1Trigger/TrackFindingTMTT/src/L1KalmanComb.cc

```diff
@@ -0,0 +1,1828 @@
+    TMatrixD hxxh = HxxH(h, pxcov);
+    tmp = dcov + hxxh;
+
+    if (getSettings()->kalmanDebugLevel() >= 4) {
+      cout << "hxxh" << endl;
+      hxxh.Print();
+      cout << "dcov + hxxh " << endl;
+      tmp.Print();
+    }
+
+    TMatrixD K(pxcovht.GetNrows(), tmp.GetNcols());
+
+    if (tmp.Determinant() == 0)
+      return K;
+    tmp.Invert();
+
+    for (int i = 0; i < pxcovht.GetNrows(); i++) {
+      for (int j = 0; j < pxcovht.GetNcols(); j++) {
+        for (int k = 0; k < tmp.GetNcols(); k++) {
+          K(i, k) += pxcovht(i, j) * tmp(j, k);
+        }
+      }
+    }
+    return K;
+  }
+
+  void L1KalmanComb::GetAdjustedState(const TMatrixD &K,
+                                      const TMatrixD &pxcov,
+                                      const std::vector<double> &x,
+                                      const StubCluster *stubCluster,
+                                      const std::vector<double> &delta,
+                                      std::vector<double> &new_x,
+                                      TMatrixD &new_xcov) const {
+    TMatrixD h = H(stubCluster);
+
+    for (int i = 0; i < K.GetNrows(); i++) {
+      new_x.at(i) = x.at(i);
+      for (int j = 0; j < K.GetNcols(); j++) {
+        new_x.at(i) += K(i, j) * delta.at(j);
+      }
+    }
+
+    TMatrixD tmp(K.GetNrows(), h.GetNcols());
+    for (int i = 0; i < K.GetNrows(); i++) {
+      tmp(i, i) = 1;
+    }
+    for (int i = 0; i < K.GetNrows(); i++) {
+      for (int j = 0; j < K.GetNcols(); j++) {
+        for (int k = 0; k < h.GetNcols(); k++) {
+          tmp(i, k) += -1 * K(i, j) * h(j, k);
+        }
+      }
+    }
+    new_xcov.Clear();
+    new_xcov.ResizeTo(pxcov.GetNrows(), pxcov.GetNcols());
+    for (int i = 0; i < tmp.GetNrows(); i++) {
+      for (int j = 0; j < tmp.GetNcols(); j++) {
+        for (int k = 0; k < pxcov.GetNcols(); k++) {
+          new_xcov(i, k) += tmp(i, j) * pxcov(j, k);
+        }
+      }
+    }
+  }
+
+  void L1KalmanComb::resetStates() {
+    for (unsigned int i = 0; i < state_list_.size(); i++) {
+      delete state_list_.at(i);
+    }
+    state_list_.clear();
+  }
+
+  const KalmanState *L1KalmanComb::mkState(const L1track3D &candidate,
+                                           unsigned skipped,
+                                           unsigned layer,
+                                           unsigned layerId,
+                                           const KalmanState *last_state,
+                                           const std::vector<double> &x,
+                                           const TMatrixD &pxx,
+                                           const TMatrixD &K,
+                                           const TMatrixD &dcov,
+                                           const StubCluster *stubCluster,
+                                           double chi2rphi,
+                                           double chi2rz) {
+    KalmanState *new_state = new KalmanState(candidate,
```

Comment:
Please use `unique_ptr` to make it clear how the ownership is moved around.

Reply:
done